flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/statistics/ContinuousHistogram.scala

```diff
@@ -0,0 +1,343 @@
+      * Find area of the trapezoid for x and the value b
+      */
+
+    val index = find(b) - 1
+    var m_b, s: Double = 0
+    if (index == -1) {
+      m_b = getCounter(index + 1) * (b - lower) / (getValue(index + 1) - lower)
+      s = m_b * (b - lower) / (2 * (getValue(index + 1) - lower))
+      return s.round.toInt
+    } else if (index == bins - 1) {
+      m_b = getCounter(index) +
+        (-getCounter(index)) * (b - getValue(index)) / (upper - getValue(index))
+      s = (getCounter(index) + m_b) *
+        (b - getValue(index)) / (2 * (upper - getValue(index)))
+    } else {
+      m_b = getCounter(index) + (getCounter(index + 1) - getCounter(index)) *
+        (b - getValue(index)) / (getValue(index + 1) - getValue(index))
+      s = (getCounter(index) + m_b) *
+        (b - getValue(index)) / (2 * (getValue(index + 1) - getValue(index)))
+    }
+    for (i <- 0 to index - 1) {
+      s = s + getCounter(i)
+    }
+    s = s + getCounter(index) / 2
+    s.round.toInt
+  }
+
+  /** Returns the total number of entries in the histogram
+    *
+    * @return Total number of points added in the histogram
+    */
+  override def total: Int = {
+    var result = 0
+    for (i <- 0 to bins - 1) {
+      result = result + getCounter(i)
+    }
+    result
+  }
+
+  /** Bin counter access function
+    *
+    * @param bin bin number to access
+    * @return `m_bin` = counter of bin
+    */
+  private[statistics] def getCounter(bin: Int): Int = {
+    require(0 <= bin && bin < bins, bin + " not in [0, " + bins + ")")
+    data.apply(bin)._2
+  }
+
+  /** Number of bins currently being used
+    *
+    * @return number of bins
+    */
+  override def bins: Int = {
+    data.size
+  }
+
+  /** Returns the string representation of the histogram.
+    *
+    */
+  override def toString: String = {
+    s"Size:" + bins + " " + data.toString
+  }
+
+  /** Loads values and counters into the histogram.
+    * This action can only be performed when there the histogram is empty
+    *
+    * @param counts Array of tuple of (value,count)
+    */
+  def loadData(counts: Array[(Double, Int)]): Unit = {
+    if (data.nonEmpty) {
+      throw new RuntimeException("Data can only be loaded during initialization")
+    }
+    counts.iterator.foreach(x => data += ((x._1, x._2)))
+    require(checkSanity, "Invalid initialization from data")
+  }
+
+  /** Bin value access function
+    *
+    * @param bin bin number to access
+    * @return `v_bin` = value of bin
+    */
+  private[statistics] def getValue(bin: Int): Double = {
+    require(0 <= bin && bin < bins, bin + " not in [0, " + bins + ")")
+    data.apply(bin)._1
+  }
+
+  /** Returns the lower limit on values of the histogram
+    *
+    * @return lower limit on values
+    */
+  private def lower: Double = {
+    min
+  }
+
+  /** Returns the upper limit on values of the histogram
+    *
+    * @return upper limit on values
+    */
+  private def upper: Double = {
+    max
+  }
```

Comment:
Because `min` and `max` is determined in initializing, `lower` and `upper` could be declared as immutable variable.